tuning/trainercontroller/controllermetrics/metrics.py

```diff
@@ -0,0 +1,218 @@
+import numpy as np
+from transformers import TrainerState, IntervalStrategy
+from transformers.utils import logging
+from collections import deque
+import math
+import copy
+
+logger = logging.get_logger(__name__)
+
+class MetricHandler:
+    """Base class for the controller-metrics"""
+    def validate(self, training_args) -> bool:
+        pass
+    
+    def compute(self, training_state, training_args=None, metrics=None) -> dict:
+        pass   
+
+class WindowStepLoss(MetricHandler):
+    """Implements the controller metric which evaluates loss-per-step over a user-defined window"""
+    
+    def __init__(self, window_size):
+        # Initialize the handler arguments
+        self.__window_size = window_size
+
+    def validate(self, training_args):
+        """Validate the training arguments (e.g logging_steps) are compatible with the computation of this metric
+
+        Args:
+            training_args: Training arguments
+
+        Returns:
+            bool
+        """
+        return training_args.logging_strategy == 'steps' and \
+            training_args.logging_steps == 1
+
+    def compute(self, training_state, training_args=None, metrics=None):
+        """Computes the controller-metric (step-loss over window) and exposes the values of the variables used by the rules.
+
+        Args:
+            training_state: TrainerState object
+            training_args: TrainingArguments object
+            metrics: [optional] metrics data
+
+        Returns:
+            dict
+        """
+        # Compute the metric using the training state
+        loss_values = [l['loss'] for l in training_state.log_history if 'loss' in l]
+        n = len(loss_values)
+        if n <= self.__window_size:
+            return None
+        window = loss_values[n-self.__window_size:n]
+        consistently_increasing = True
+        for i in range(len(window)-1):
+            if window[i] > window[i+1]:
+                consistently_increasing = False
+                break
+        w = np.array(window)
+        avg_loss = np.mean(w)
+        std_loss = np.std(w, dtype=np.float64)
+        exposed_data = {'consistently_increasing': int(consistently_increasing), \
+                    'average_loss': avg_loss,\
+                    'std_loss': std_loss,\
+                    'window': window}
```

Comment:
Should we call it something like, loss_in_window?

Reply:
window seems better... `loss_in_window` can be inferred by looking at class name and argument name.